src/plugins/FileService/arweave/index.ts

```diff
@@ -0,0 +1,101 @@
+import { Attachment } from '@dimensiondev/common-protocols'
+import Arweave from 'arweave/web'
+import type Transaction from 'arweave/web/lib/transaction'
+import { landing } from '../constants'
+import { sign } from './remote-signing'
+import token from './token.json'
+import { isNil, isEmpty } from 'lodash-es'
+
+const stage: Record<Transaction['id'], Transaction> = {}
+
+const instance = Arweave.init({
+    host: 'arweave.net',
+    port: 443,
+    protocol: 'https',
+})
+
+export function makeFileKey(length = 16) {
+    let key = ''
+    const table = 'ABDEFGHJKMNPQRTWXYadefhijkmnprstuvwxyz03478'
+    for (let i = 0; i < length; i += 1) {
+        key += table.charAt(Math.floor(Math.random() * table.length))
+    }
+    return key
+}
+
+export interface AttachmentOptions {
+    key?: string | null
+    type: string
+    block: Uint8Array
+}
+
+export async function makeAttachment(options: AttachmentOptions) {
+    const passphrase = options.key ? encodeText(options.key) : undefined
+    const encoded = await Attachment.encode(passphrase, {
+        block: options.block,
+        metadata: null,
+        mime: isEmpty(options.type) ? 'application/octet-stream' : options.type,
+    })
+    const transaction = await makePayload(encoded, 'application/octet-stream')
+    stage[transaction.id] = transaction
+    return transaction.id
+}
+
+// import { ServicesWithProgress } from 'src/extension/service.ts'
+// ServicesWithProgress.pluginArweaveUpload
+export async function* upload(id: Transaction['id']) {
+    for await (const uploader of instance.transactions.upload(stage[id])) {
```

Comment:
Please notice that the async generator is powered by the caller. If the caller is no longer calling `.next()` on the iterator, the process won't go on. Please make sure the uploading process will continue even the web page is closed (if it's the expected behavior).

Reply:
yep, this upload cannot be cancelled